server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -835,37 +891,69 @@ private long appendModColumns(final View view, final long startRange, final int
             final Consumer<InputStream> addStream,
             final ChunkInputStreamGenerator.FieldNodeListener fieldNodeListener,
             final ChunkInputStreamGenerator.BufferListener bufferListener) throws IOException {
-        long endRange = startRange + targetBatchSize;
         int[] columnChunkIdx = new int[modColumnData.length];
+
         // for each column identify the chunk that holds this startRange
+        long maxLength = targetBatchSize;
+
+        // adjust the batch size if we would cross a chunk boundary
         for (int ii = 0; ii < modColumnData.length; ++ii) {
             final ModColumnData mcd = modColumnData[ii];
-            int chunkIdx = (int) (startRange / SNAPSHOT_CHUNK_SIZE);
-            // adjust the batch size if we would cross a chunk boundary
-            endRange = Math.min((long) (chunkIdx + 1) * SNAPSHOT_CHUNK_SIZE, endRange);
+            final ChunkInputStreamGenerator[] generators = mcd.data.generators;
+            int chunkIdx = 0;
+            if (generators.length > 0) {
+                final RowSet modOffsets = view.modRowOffsets(ii);
+                // if all mods are being sent, then offsets yield an identity mapping
+                final long startPos = modOffsets != null ? modOffsets.get(startRange) : startRange;
+                if (startPos != RowSet.NULL_ROW_KEY && chunkIdx < generators.length - 1) {
+                    chunkIdx = findGeneratorForOffset(generators, startPos);
+                    maxLength = Math.min(maxLength, generators[chunkIdx].getLastRowOffset() + 1 - startPos);
```

Comment:
This doesn't make sense to me. Maybe `maxLength` was initialized wrong, and we want a `Math.max` here?